src/v/rpc/connection_cache.cc

```diff
@@ -61,11 +61,14 @@ ss::future<> connection_cache::remove(model::node_id n) {
 /// \brief closes all client connections
 ss::future<> connection_cache::stop() {
     auto units = co_await _mutex.get_units();
-    co_await parallel_for_each(_cache, [](auto& it) {
+    // Exchange ensures the cache is invalidated and concurrent
+    // accesses wait on the mutex to populate new entries.
+    auto cache = std::exchange(_cache, {});
+    co_await parallel_for_each(cache, [](auto& it) {
         auto& [_, cli] = it;
         return cli->stop();
     });
-    _cache.clear();
+    cache.clear();
```

Comment:
the cache is now going to go out of scope on the next line and cleared in any case